features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/vitals/JankStatsActivityLifecycleListener.kt

```diff
@@ -200,20 +173,90 @@ internal class JankStatsActivityLifecycleListener(
+            if (jankStats == null) {
+                internalLogger.log(
+                    InternalLogger.Level.WARN,
+                    InternalLogger.Target.MAINTAINER,
+                    { "Unable to create JankStats" }
+                )
+            } else {
+                activeWindowsListener[window] = jankStats
+            }
+        }
+    }
+
+    @SuppressLint("NewApi")
+    private fun trackWindowMetrics(isKnownWindow: Boolean, window: Window, activity: Activity) {
+        if (buildSdkVersionProvider.version() >= Build.VERSION_CODES.S && !isKnownWindow) {
+            registerMetricListener(window)
+        } else if (display == null && buildSdkVersionProvider.version() == Build.VERSION_CODES.R) {
+            // Fallback - Android 30 allows apps to not run at a fixed 60hz, but didn't yet have
+            // Frame Metrics callbacks available
+            val displayManager = activity.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager
+            display = displayManager.getDisplay(Display.DEFAULT_DISPLAY)
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.S)
     private fun registerMetricListener(window: Window) {
         if (buildSdkVersionProvider.version() >= Build.VERSION_CODES.S) {
             if (frameMetricsListener == null) {
                 frameMetricsListener = DDFrameMetricsListener()
             }
             val handler = Handler(Looper.getMainLooper())
-            frameMetricsListener?.let { window.addOnFrameMetricsAvailableListener(it, handler) }
+            // Only hardware accelerated views can be tracked with metrics listener
+            if (window.peekDecorView()?.isHardwareAccelerated == true) {
+                frameMetricsListener?.let { listener ->
+                    try {
+                        @Suppress("UnsafeThirdPartyFunctionCall") // Listener can't be null here
+                        window.addOnFrameMetricsAvailableListener(listener, handler)
+                    } catch (e: IllegalStateException) {
+                        internalLogger.log(
+                            InternalLogger.Level.ERROR,
+                            InternalLogger.Target.MAINTAINER,
+                            { "Unable to attach JankStatsListener to window" },
+                            e
+                        )
+                    }
+                }
+            } else {
+                internalLogger.log(
+                    InternalLogger.Level.WARN,
+                    InternalLogger.Target.MAINTAINER,
+                    { "Unable to attach JankStatsListener to window, decorView is null or not hardware accelerated" }
+                )
+            }
         }
     }
 
     @RequiresApi(Build.VERSION_CODES.N)
     private fun unregisterMetricListener(window: Window) {
-        window.removeOnFrameMetricsAvailableListener(frameMetricsListener)
+        try {
+            window.removeOnFrameMetricsAvailableListener(frameMetricsListener)
```

Comment:
It is probably worth to check if `frameMetricsListener` is not null here before calling the method. Normally it cannot be at this point, since we got `onStart` call and created it, but just in case if it breaks in the future due to some refactoring.

In theory it is worth to check here for the hardware-accelerated view as well, but only if we are sure that `peekDecorView` at this point cannot be null (otherwise we may fail to remove listener).